### ISSUE TITLE
trait bound is reported not satisfied although it is

### ISSUE THREAD

**Issue by sidgilles** [rust-lang/rust#61021]
Hello!

Rustc report an unsatisfied trait bound. However, following the compiler suggestion does not help much. Working with tokio.
Is #43580 also relevant for this one ?
Thanks.

Sample code.
```rust

use tokio::codec::{Encoder, Decoder, LinesCodec, Framed};
use tokio::io::{AsyncRead, AsyncWrite, Error};
// other uses here...

struct DummyServer<T, U>{
    pub stream_instant : Box<Stream<Item=Instant, Error=tokio::timer::Error> + Send>,
    pub stream_line : Box<Framed<T, U>>,
    pub closure_timer : Box<FnMut(bool, &mut Box<Framed<T, U>>, Option<<U as Decoder>::Item>) -> () + Send>,
}

impl <T, U> Future for DummyServer<T, U>
    where U : Encoder + Decoder, T : AsyncRead + AsyncWrite
 {
    type Item = ();
    type Error = ();

    fn poll(&mut self) ->  Result<Async<Self::Item>, Self::Error> {
       // some code here...
    }
}
```

Compiler output.
```
error[E0277]: the trait bound `U: tokio::codec::Decoder` is not satisfied
  --> driver_com/common/src/main.rs:54:5
   |
54 |     pub closure_timer : Box<FnMut(bool, &mut Box<Framed<T, U>>, Option<<U as Decoder>::Item>) -> () + Send>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `tokio::codec::Decoder` is not implemented for `U`
   |
   = help: consider adding a `where U: tokio::codec::Decoder` bound
```

This is happening both on stable and nightly, freshly updated.
```
  stable-x86_64-unknown-linux-gnu unchanged - rustc 1.34.2 (6c2484dc3 2019-05-13)
   nightly-x86_64-unknown-linux-gnu updated - rustc 1.36.0-nightly (d35181ad8 2019-05-20)
```

Dependencies.
```
[dependencies]
futures = { package = "futures", version = "0.1" }
tokio = { version = "0.1" }
```

**Edit:**
This sample code get compiled.
I would say that:
  1) the message `help: consider adding a 'where U: tokio::codec::Decoder' bound` is misleading as the `where` clause does not apply on struct definition
  2) it would be nice to see this unusual case in the docs, really nice.

```rust
struct DummyServer<T, U : Encoder + Decoder>{
    pub stream_instant : Box<Stream<Item=Instant, Error=tokio::timer::Error> + Send>,
    pub stream_line : Box<Framed<T, U>>,
    pub closure_timer : Box<FnMut(bool, &mut Box<Framed<T, U>>, Option<<U as Decoder>::Item>) -> () + Send>,
}

impl <T,U> Future for DummyServer<T,U>
    where T : AsyncRead + AsyncWrite, U : Encoder + Decoder
 {
    type Item = ();
    type Error = ();

    fn poll(&mut self) ->  Result<Async<Self::Item>, Self::Error> {
       // some code here...
    }
}
```

**Comment by apiraino** [comment]
Closing as an MCVE could not be provided. The provided sample code does not emit the same error anymore, I've tried to compile with rustc 1.34.2 but failed (I guess it makes it even trickier to reproduce today) due to external dependencies not supporting it anymore?
